contributors/design-proposals/api-machinery/customresources-versioning.md

```diff
@@ -0,0 +1,177 @@
+type CustomResourceDefinitionStatus struct {
+  ...
+
+  // StoredVersions are all versions ever marked as storage in spec. Tracking these
+  // versions allow a migration path for stored version in etcd. The field is mutable
+  // so the migration controller can first make sure a version is certified (i.e. all
+  // stored objects is that version) then remove the rest of the versions from this list.
+  // None of the versions in this list can be removed from the spec.Versions field.
+  StoredVersions []string
+}
+```
+
+Basic validations needed for the `version` field are:
+
+* `Spec.Version` field exists in `Spec.Versions` field (serve should be True for that version if there is any).
+* If there is any `Served` `Version` in list of `Versions`, then the `Spec.Version`’s version in the `Spec.Versions` list should have a set `Served` flag. This is for backward compatibility. An old controller expect that version to be served but only the whole CRD is served. CRD Registration controller should unregister a CRD with no serving version.
+* None of the `Status.StoredVersion` can be removed from `Spec.Versions` list.
+* Only one of the versions in `spec.Versions` can flag as `Storage` version.
+
+# **Declarative conversion**
+
+Most of CRD conversions can be represented by simple operations like rename or move. The proposal is to support rename only and add more operations in future based on user feedback.
+
+```golang
+type CustomResourceDefinitionVersion {
+  // Name is the version name, e.g. "v1", “v2beta1”, etc.
+  Name string
+  // Served is a flag enabling/disabling this version from being served via REST APIs
+  Served Boolean
+  // Storage flags the release as a storage version. There can be only one version
+  // flagged as Storage.
+  Storage Boolean
+  //
+  Conversion CustomResourceConversion
+}
+
+type CustomResourceConversion struct {
+  Declarative map[string]CustomResourceDeclarativeConversion
+}
+
+type CustomResourceDeclarativeConversion {
+  // This type will have collection of conversion operations. Only one of them should be set.
+  Rename RenameConversion
+}
+
+type RenameConversion {
+  From JsonPath
+  To JsonPath
+}
+
+type JsonPath string
+```
+
+# **External Conversion/Webhooks**
+
+Webhooks gives flexibility to CRD author to convert from the specific version to any version. As the conversion needs to be reversible, here we only list versions this webhook converts to/from. 
+
+```golang
+type CustomResourceDefinitionVersion {
+  // Name is the version name, e.g. "v1", “v2beta1”, etc.
+  Name string
+  // Served is a flag enabling/disabling this version from being served via REST APIs
+  Served Boolean
+  // Storage flags the release as a storage version. There can be only one version
+  // flagged as Storage.
+  Storage Boolean
+  //
+  Conversion CustomResourceConversion
+}
+
+type CustomResourceConversion struct {
+  Declarative map[string]CustomResourceDeclarativeConversion
+  External CustomResourceConversionWebhook
```

Comment:
pointer and omitempty

Reply:
Having a webhook definition per version looks odd. I would expect it globally with a set of support conversions pairs. Then you don't have to maintain `n` webhook structs. Alternatively, we could think about a global named conversion webhook field and reference the names here (like volumes and mounts in pods).

Reply:
I explained the design further down. It is cleaner to define the webhook on the version but the validation will not let user define more than one webhook. It will just make sure if more than one webhook is defined, they are the same.

Reply:
Don't like this kind of redundancy. Why is it cleaner?